src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -309,51 +358,55 @@ protected abstract AbstractLdapNamingEnumeration<? extends NameClassPair> getRef
+        try {
+            if ((refEx != null) &&
+                (refEx.hasMoreReferrals() ||
+                 refEx.hasMoreReferralExceptions()
+                    && !(errEx instanceof LimitExceededException))) {
 
-        if ((refEx != null) &&
-            (refEx.hasMoreReferrals() ||
-             refEx.hasMoreReferralExceptions()
-                && !(errEx instanceof LimitExceededException))) {
+                if (homeCtx().handleReferrals == LdapClient.LDAP_REF_THROW) {
+                    throw (NamingException)(refEx.fillInStackTrace());
+                }
 
-            if (homeCtx.handleReferrals == LdapClient.LDAP_REF_THROW) {
-                throw (NamingException)(refEx.fillInStackTrace());
-            }
+                // process the referrals sequentially
+                while (true) {
 
-            // process the referrals sequentially
-            while (true) {
+                    LdapReferralContext refCtx =
+                        (LdapReferralContext)refEx.getReferralContext(
+                        homeCtx().envprops, homeCtx().reqCtls);
 
-                LdapReferralContext refCtx =
-                    (LdapReferralContext)refEx.getReferralContext(
-                    homeCtx.envprops, homeCtx.reqCtls);
+                    try {
 
-                try {
+                        update(getReferredResults(refCtx));
+                        break;
 
-                    update(getReferredResults(refCtx));
-                    break;
+                    } catch (LdapReferralException re) {
 
-                } catch (LdapReferralException re) {
+                        // record a previous exception
+                        if (errEx == null) {
+                            errEx = re.getNamingException();
+                        }
+                        refEx = re;
+                        continue;
 
-                    // record a previous exception
-                    if (errEx == null) {
-                        errEx = re.getNamingException();
+                    } finally {
+                        // Make sure we close referral context
+                        refCtx.close();
                     }
-                    refEx = re;
-                    continue;
-
-                } finally {
-                    // Make sure we close referral context
-                    refCtx.close();
                 }
-            }
-            return hasMoreImpl();
+                return hasMoreImpl();
 
-        } else {
-            cleanup();
+            } else {
+                cleanup();
 
-            if (errEx != null) {
-                throw errEx;
+                if (errEx != null) {
+                    throw errEx;
+                }
+                return (false);
             }
-            return (false);
+        } finally {
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```

Comment:
> Ditto, the try/finally is unnecessary.

I think it is necessary for reachability fence since try block has many exit paths and accesses state cleaned up by Cleaner. Memory fence OTOH is not needed.